### PR TITLE
Integrate tailscale into k3s

### PR DESCRIPTION
**Pull request by manuelbuil**
#### Proposed Changes ####

This PR adds the necessary code to integrate VPNs into k3s and incorporates tailscale as the first VPN option. 
It also includes an ADR that explains the change

#### Types of Changes ####

New feature

#### Verification ####

1 - Install tailscale in the node
2 - Create a key in tailscale
3 - Deploy k3s with the flag:
```
vpn-auth: "name=tailscale,joinKey=$CREATED_KEY"
```
4 - Verify that:
a) Tailscale interface exists and has an IP
b) `kubectl get endpoints` is using the tailscale IP
c) `kubectl get nodes -o wide` shows the tailscale IP as the nodeIP

#### Linked Issues ####

https://github.com/k3s-io/k3s/issues/7353

#### User-Facing Change ####

```release-note
Integration of tailscale VPN into k3s
```

### CODE REVIEW

File: pkg/agent/flannel/setup.go
Comment (brandond) [inline pkg/agent/flannel/setup.go:73]: user is expected to provide tailscale binaries on their own, I take it?

File: pkg/agent/flannel/setup.go
Comment (brandond) [inline pkg/agent/flannel/setup.go:93]: user is expected to provide curl and netclient binaries on their own? Note that our k3s-root provides only a busybox curl stub that doesn't support TLS or most of the flags you're using here.

File: pkg/agent/flannel/setup.go
Comment (manuelbuil) [reply pkg/agent/flannel/setup.go:73]: Yes! That will be explained in the docs. It is explained in the ADR: "At this stage, the proposal is not to incorporate the vpn binaries or daemons inside the k3s binary, we require the user to have the vpn binary installed."

And also important: "The collection of vpn endpoint information and its start are implemented by calling the vpn binary. Tailscale has a "not-feature-complete" go client but netmaker does not, so calls to the binary is the common denominator"

**Review by dereknola — CHANGES_REQUESTED** [review]

File: pkg/netutil/vpn.go
Comment (dereknola) [reply pkg/netutil/vpn.go:130]: In fact looking at the whole package, I think `netutil` should either be integrated into `util` OR some of the files in `util`, like `net.go` so be moved under `netutil`. I know this is somewhat tangential to your VPN work, that stuff looks great :+1:

Comment (manuelbuil) [reply pkg/netutil/vpn.go:130]: good point! I never really noticed we have two different "utils" dirs: netutils and utils. I guess we should consolidate everything into just one (probably utils). If nobody picks it up, I could create a different PR tomorrow doing that change

File: pkg/agent/flannel/setup.go
Comment (matttrach) [reply pkg/agent/flannel/setup.go]: What is `api.mbuilnetmaker.ydns.eu`?

Comment (manuelbuil) [reply pkg/agent/flannel/setup.go]: That was part of the old code including netmaker stuff. I removed it and just left tailscale :). Have a look at the update from 2h ago, there shouldn't be anything saying netmaker (I hope)

**Review by dereknola — APPROVED** [review]

File: pkg/cli/cmds/agent.go
Comment (brandond) [inline pkg/cli/cmds/agent.go:155]: I think this needs to 
* Be handled by the code that redacts secrets from the node-args annotation
* Have an environment variable so that it can be set without a CLI flag
* Maybe support reading from a file? I can already see the issues asking for this...

File: pkg/netutil/vpn.go
Comment (brandond) [inline pkg/netutil/vpn.go:23]: The VPN prefix on this stutters. Do you think all vpns will have a key, or should we have key+secret or user+password?
```suggestion
	APIKey string
```

File: pkg/netutil/vpn.go
Comment (brandond) [inline pkg/netutil/vpn.go:30]: Is this a copy of VPNInfo's APIKey?

File: pkg/netutil/vpn.go
Comment (brandond) [inline pkg/netutil/vpn.go:40]: This should probably be a switch/default even though we only support one VPN.

**Review by brandond — DISMISSED** [review]

File: pkg/netutil/vpn.go
Comment (manuelbuil) [reply pkg/netutil/vpn.go:23]: All VPNs have a JoinKey, which allows clients to register into the VPN mesh. APIKey is something required specifically for netmaker because the binary is not feature complete and you need to `curl` into the server for certain configs. You know what, let's remove that APIKey and if we ever integrate netmaker (or any other VPN that needs the workaround) we will add it

File: pkg/netutil/vpn.go
Comment (manuelbuil) [reply pkg/netutil/vpn.go:30]: Yes, it was. I will remove APIKey

File: pkg/netutil/vpn.go
Comment (manuelbuil) [reply pkg/netutil/vpn.go:40]: make sense

File: pkg/cli/cmds/agent.go
Comment (manuelbuil) [reply pkg/cli/cmds/agent.go:155]: Regarding the third point, are you suggesting to create a new cli flag which points at a file that contains the vpn-auth info?

**Comment by cwayne18** [comment]
Can we get either an integration or an e2e test for this?

File: pkg/cli/cmds/agent.go
Comment (brandond) [reply pkg/cli/cmds/agent.go:155]: Yes, similar to what we do for the token (--token-file)

**Comment by nikolaishields** [comment]
From a community perspective it would be really nice to include the ability to setup a [tailscale funnel](https://tailscale.com/kb/1223/tailscale-funnel/) at node deploy time to allow public access to the node at a targeted port or service address. This would come in particularly convenient in situations where public ingress is desired. A few example use-cases:
- IDP (keycloak, vault, ory, dex) is deployed to the cluster by the k3s helm-manifest controller at cluster deploy. This would allow k3s users to have SSO access to their clusters immediately after bootstrap.
- Traefik ingress is proxying public services, yet the cluster is behind CGNAT and ports cannot be exposed.
- HTTP01 cert-generation via public exposure of the ACME endpoint
Perhaps a Kubernetes ingress integration could even be designed that instantiates a funnel as a CRD when ingress is created. The sky's the limit. 
Special thanks to @manuelbuil for drafting this as it looks awesome!

**Comment by manuelbuil** [comment]
> From a community perspective it would be really nice to include the ability to setup a [tailscale funnel](https://tailscale.com/kb/1223/tailscale-funnel/) at node deploy time to allow public access to the node at a targeted port or service address. This would come in particularly convenient in situations where public ingress is desired. A few example use-cases:
> 
> * IDP (keycloak, vault, ory, dex) is deployed to the cluster by the k3s helm-manifest controller at cluster deploy. This would allow k3s users to have SSO access to their clusters immediately after bootstrap.
> * Traefik ingress is proxying public services, yet the cluster is behind CGNAT and ports cannot be exposed.
> * HTTP01 cert-generation via public exposure of the ACME endpoint
> 
> Perhaps a Kubernetes ingress integration could even be designed that instantiates a funnel as a CRD when ingress is created. The sky's the limit. Special thanks to @manuelbuil for drafting this as it looks awesome!

Hey @nikolaishields, how are you? Thanks for your suggestion :). There are several cool features that tailscale provides, such as funneling, but I decided to keep them outside of the integration (at least for the time being). I want to keep it dead simple and check what is the community reaction, typical bugs, etc before taking the next iteration.

The CRD+ingress idea could be cool. I guess you mean having the ingress service as a nodePort service and offering that port with tailscale funnel. Kind of a "type: loadbalancer" service but without the loadbalancer, right? Wouldn't klipper-lb fulfill that use case too? Now I want/need to test it

**Comment by manuelbuil** [comment]
> Can we get either an integration or an e2e test for this?

Yep, on it

**Comment by nikolaishields** [comment]
>Hey @nikolaishields, how are you?

I'm doing well thanks for asking!
>There are several cool features that tailscale provides, such as funneling, but I decided to keep them outside of the integration (at least for the time being). I want to keep it dead simple and check what is the community reaction, typical bugs, etc before taking the next iteration.

Makes perfect sense, if you ever want/need a hand on implementing any of these features let me know. Happy to help!
>The CRD+ingress idea could be cool. I guess you mean having the ingress service as a nodePort service and offering that port with tailscale funnel. Kind of a "type: loadbalancer" service but without the loadbalancer, right? Wouldn't klipper-lb fulfill that use case too? Now I want/need to test it

It probably could and then a user could just run a funnel against the node port. In theory I suppose one could also just funnel a service directly via local dns lookup (e.g. `tailscale serve <servicename>.default.cluster.local:<port number>`).

**Comment by manuelbuil** [comment]
> Can we get either an integration or an e2e test for this?

Actually, could we merge this PR and then work on the e2e test? It complicates things that the code is not yet part of upstream k3s

**Comment by brandond** [comment]
I would defer to @cwayne18 on merging without tests. I think @dereknola might have some suggestions on how to add tests before the PR is merged? I have had some frustrations with that process as well.

File: tests/e2e/tailscale/tailscale_test.go
Comment (dereknola) [inline tests/e2e/tailscale/tailscale_test.go:18]: nit: you don't use/need this argument for this test.

File: pkg/cli/cmds/agent.go
Comment (brandond) [inline pkg/cli/cmds/agent.go:164]:
```suggestion
		Usage:       "(agent/networking) File containing credentials for the VPN provider. It must include the provider name and join key in the format name=foo,joinKey=var",
```

File: pkg/cli/cmds/agent.go
Comment (brandond) [inline pkg/cli/cmds/agent.go:158]:
```suggestion
		Usage:       "(agent/networking) Credentials for the VPN provider. It must include the provider name and join key in the format name=foo,joinKey=var",
```

File: pkg/agent/config/config.go
Comment (brandond) [inline pkg/agent/config/config.go:397]:
```suggestion
				logrus.Warn("VPN provider overrides configured node-external-ip parameter")
```

File: pkg/agent/config/config.go
Comment (brandond) [inline pkg/agent/config/config.go:394]:
```suggestion
				logrus.Warn("VPN provider overrides configured node-ip parameter")
```

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/util/vpn.go
Comment (briandowns) [inline pkg/util/vpn.go:109]: What happens to the error?

File: pkg/util/vpn.go
Comment (briandowns) [inline pkg/util/vpn.go:47]:
```suggestion
		return errors.New("Requested VPN: " + authInfo.Name + " is not supported. We currently only support tailscale")
```

Generally we should avoid using formatted strings with strings as it's a lot of unnecessary memory/heap usage due ot the use of reflection with no benefit.

File: pkg/util/vpn.go
Comment (briandowns) [inline pkg/util/vpn.go:1]: We can probably create a "vpn" package and not have random stuff added to the "utils" package.

File: pkg/agent/config/config.go
Comment (manuelbuil) [reply pkg/agent/config/config.go:559]: What would be the benefit of doing this? It means adding a new function and two extra lines of code

File: pkg/agent/flannel/setup.go
Comment (rbrtbnfgl) [inline pkg/agent/flannel/setup.go:250]: Why do you need to do this again?

File: pkg/cli/cmds/agent.go
Comment (rbrtbnfgl) [inline pkg/cli/cmds/agent.go:168]: The `&` was removed here.

File: pkg/daemons/config/types.go
Comment (rbrtbnfgl) [inline pkg/daemons/config/types.go:28]: Is it needed to update the `flannel-backend` flag info with the new type of backend? Is it possible to configure tailscale as backend without `vpn-auth` definition?

File: pkg/agent/flannel/setup.go
Comment (manuelbuil) [reply pkg/agent/flannel/setup.go:250]: Good catch, it is not needed. Probably a copy/paste mistake

File: pkg/cli/cmds/agent.go
Comment (manuelbuil) [reply pkg/cli/cmds/agent.go:168]: Thanks! This is a mistake

File: pkg/daemons/config/types.go
Comment (manuelbuil) [reply pkg/daemons/config/types.go:28]: If the user connects the local tailscale instance to the tailscale server (by logging in), then things will work if user passes tailscale as the flannel-backend. I am not sure if we want to document that, it will probably create confusion

File: pkg/cli/cmds/agent.go
Comment (rbrtbnfgl) [inline pkg/cli/cmds/agent.go:158]: Maybe using valid value could it be more understandable `name=tailscale,joinKey=<key>`

Comment (manuelbuil) [reply pkg/cli/cmds/agent.go:158]: yeah, I can do that. `name=tailscale` might be good right now but if we add more vpns, then it may confuse people. Something to consider in the future

Comment (brandond) [reply pkg/cli/cmds/agent.go]: even just `name=PROVIDER,joinKey=MYKEY` might be more clear?

Comment (rbrtbnfgl) [reply pkg/cli/cmds/agent.go]: Maybe if we decided to mark this as experimental could be better to write it on the flag description?

Comment (manuelbuil) [reply pkg/cli/cmds/agent.go]: Done! :)

**Review by rbrtbnfgl — APPROVED** [review]

File: pkg/cli/cmds/agent.go
Comment (rbrtbnfgl) [reply pkg/cli/cmds/agent.go]: I think that also `vpn-auth-file` should be marked as experimental

Comment (manuelbuil) [reply pkg/cli/cmds/agent.go]: yeah, good catch

File: pkg/agent/config/config.go
Comment (manuelbuil) [reply pkg/agent/config/config.go:564]: Got you, you convinced me. I added two extra fields to VPNInfo:
```
type VPNInfo struct {
	IPs          []net.IP
	NodeID       string
	ProviderName string
	VPNInterface string
}
```

`vpnInfo.ProviderName` is passed directly to `nodeConfig.FlannelBackend` and `vpnInfo.VPNInterface` is passed directly to `flannelIface, err = net.InterfaceByName(vpnInfo.VPNInterface)`. There is nothing saying tailscale in `config/config.go`

Comment (manuelbuil) [reply pkg/agent/config/config.go:564]: And we don't need to query GetVPNInfo again because we are doing it already earlier to set the nodeIP and flannelIface

File: pkg/util/command.go
Comment (dereknola) [inline pkg/util/command.go:17]: With this, you get more info if there is an error, the err outs from exec are almost always just `exit 1/2`, which isn't helpful.
```suggestion
	var out, errOut bytes.Buffer

	cmd := exec.Command(command, args...)
	cmd.Stdout = &out
	cmd.Stderr = &errOut
	err := cmd.Run()
	if err != nil {
		return errOut.String(), err
	}
```

File: pkg/util/file.go
Comment (dereknola) [inline pkg/util/file.go:27]: I would suggest adding a failure state is after 5min or something, the path still does not exist.

File: pkg/util/file_windows.go
Comment (dereknola) [inline pkg/util/file_windows.go:19]: Same comment as in the linux version

File: tests/e2e/tailscale/Vagrantfile
Comment (dereknola) [inline tests/e2e/tailscale/Vagrantfile:23]: Do we need the special libvirt flags? Its not dualstack
```suggestion
 vm.network "private_network", ip: node_ip, netmask: "255.255.255.0"
```

File: docs/adrs/integrat-vpns.md
Comment (brandond) [inline docs/adrs/integrat-vpns.md:1]: nit on the filename - should it be `integrate-vpns.md`? I'm not sure why the word was truncated.

File: pkg/cli/cmds/agent.go
Comment (brandond) [inline pkg/cli/cmds/agent.go:158]: nit: be consistent with the format for `vpn-auth-file`
```suggestion
		Usage:       "(agent/networking) (experimental) Credentials for the VPN provider. It must include the provider name and join key in the format name=<vpn-provider>,joinKey=<key>",
```

File: pkg/util/file.go
Comment (brandond) [inline pkg/util/file.go:22]: I know that this is an existing function that you moved, but I don't like that this infinitely retries on nonexistent files. It should either have a timeout, or accept a context, or both. The function name and docstring should also indicate that it retries at intervals if the file doesn't exist.

For an example of how this might be problematic - if the user points k3s at a nonexistent vpn token file, k3s will just hang forever on startup, when in reality we probably just want to fail immediately if the file doesn't exist. This function makes more sense for its existing use in the codebase, where it's generally used to get content from a file that is dropped by code that is running in parallel to the goroutine trying to read it.

Take a look at `wait.PollImmediateWithContext` elsewhere in the codebase, that's probably what I'd use here.

File: pkg/util/file_windows.go
Comment (brandond) [reply pkg/util/file_windows.go:19]: I don't understand why this is OS-specific, there doesn't appear to be anything in either function that would differ across platforms.

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/util/command.go
Comment (manuelbuil) [reply pkg/util/command.go:17]: nice, thanks for that. The user of this function should then be aware that the interesting error is the string returned and not the error, right? I'll add a comment

File: pkg/util/file.go
Comment (manuelbuil) [reply pkg/util/file.go:27]: yeah, good point, let me replace the infinite loop with something else using time. I copied/pasted this function from another place though, and it never happened before, but better safe than sorry

File: pkg/util/file_windows.go
Comment (manuelbuil) [reply pkg/util/file_windows.go:19]: We could remove it in another PR (just in case we need to revert) and see if things are still working

File: tests/e2e/tailscale/Vagrantfile
Comment (manuelbuil) [reply tests/e2e/tailscale/Vagrantfile:23]: oh, I think we are using those flags in all the networking tests. We should remove it when not needed then. Thanks!

File: docs/adrs/integrat-vpns.md
Comment (manuelbuil) [reply docs/adrs/integrat-vpns.md:1]: good catch, probably my mistake

File: pkg/cli/cmds/agent.go
Comment (manuelbuil) [reply pkg/cli/cmds/agent.go:158]: :+1:

File: pkg/util/file.go
Comment (manuelbuil) [reply pkg/util/file.go:22]: Derek had the same comment, I added a timeout after 4 minutes

**Review by briandowns — APPROVED** [review]

**Review by dereknola — APPROVED** [review]

**Pull request merged**
